fhir-cache/src/main/java/com/ibm/fhir/cache/CacheManager.java

```diff
@@ -0,0 +1,306 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.cache;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.ibm.fhir.cache.util.CacheSupport;
+import com.ibm.fhir.core.TenantIdProvider;
+
+/**
+ * A class used to create and manage cache instances on a per tenant basis
+ */
+public final class CacheManager {
+    private static final Map<String, Map<String, Cache<?, ?>>> TENANT_CACHE_MAPS = new ConcurrentHashMap<>();
+    private static final TenantIdProvider TENANT_ID_PROVIDER = TenantIdProvider.provider();
+
+    private CacheManager() { }
+
+    /**
+     * Get the cache names for the current tenant.
+     *
+     * @return
+     *     the cache names
+     */
+    public Set<String> getCacheNames() {
+        String tenantId = TENANT_ID_PROVIDER.getTenantId();
+        return TENANT_CACHE_MAPS.getOrDefault(tenantId, Collections.emptyMap()).keySet();
+    }
+
+    /**
+     * Get the managed cache with the given name for the current tenant.
+     *
+     * @param <K>
+     *     the key type
+     * @param <V>
+     *     the value type
+     * @param cacheName
+     *     the cache name
+     * @return
+     *     the managed cache instance with the given name, or null if not exists
+     */
+    @SuppressWarnings("unchecked")
```

Comment:
I assume there is no way to avoid this...
Is the cast always safe?

Reply:
Unfortunately, the cast is unavoidable unless we want consumers of the API to handle it which seems ugly when we can do it in this code. The cast is not always safe. `javax.cache.CacheManager` does something similar.